model/test_results.go

```diff
@@ -579,6 +586,117 @@ func FindTestResults(ctx context.Context, env cedar.Environment, opts FindTestRe
 	return results, nil
 }
 
+// FindTestSamplesOptions specifies test results to collect test names from
+// and regex filters to apply to the test names.
+type FindTestSamplesOptions struct {
+	Tasks           []FindTestResultsOptions
+	TestNameRegexes []string
+}
+
+func (opts *FindTestSamplesOptions) createFindQuery() map[string]interface{} {
+	findQueries := []bson.M{}
+	for _, task := range opts.Tasks {
+		findQueries = append(findQueries, task.createFindQuery())
+	}
+
+	return bson.M{"$or": findQueries}
+}
+
+func (opts *FindTestSamplesOptions) createFindOptions() *options.FindOptions {
+	return options.Find().SetProjection(bson.M{
+		bsonutil.GetDottedKeyName(testResultsInfoKey, testResultsInfoTaskIDKey):        1,
+		bsonutil.GetDottedKeyName(testResultsInfoKey, testResultsInfoDisplayTaskIDKey): 1,
+		bsonutil.GetDottedKeyName(testResultsInfoKey, testResultsInfoExecutionKey):     1,
+		testResultsFailedTestsSampleKey:                                                1,
+	})
+}
+
+func (opts *FindTestSamplesOptions) makeTestSamples(testResults []TestResults) []TestResultsSample {
+	samples := consolidateSamples(testResults)
+	if len(opts.TestNameRegexes) > 0 {
+		samples = filterTestNames(samples, opts.TestNameRegexes)
+	}
+
+	return samples
+}
+
+func consolidateSamples(testResults []TestResults) []TestResultsSample {
+	type taskExecutionPair struct {
+		taskID    string
+		execution int
+	}
+	taskPairMap := make(map[taskExecutionPair][]string)
+	for _, result := range testResults {
+		taskID := result.Info.DisplayTaskID
+		if taskID == "" {
+			taskID = result.Info.TaskID
+		}
+
+		pair := taskExecutionPair{
+			taskID:    taskID,
+			execution: result.Info.Execution,
+		}
+		taskPairMap[pair] = append(taskPairMap[pair], result.FailedTestsSample...)
+	}
+
+	samples := make([]TestResultsSample, 0, len(taskPairMap))
+	for pair, names := range taskPairMap {
+		samples = append(samples, TestResultsSample{
+			TaskID:          pair.taskID,
+			Execution:       pair.execution,
+			FailedTestNames: names,
+		})
+	}
+
+	return samples
+}
+
+func filterTestNames(samples []TestResultsSample, regexStrings []string) []TestResultsSample {
+	regexes := []*regexp.Regexp{}
+	for _, regexString := range regexStrings {
+		testNameRegex, err := regexp.Compile(regexString)
+		if err != nil {
+			continue
```

Comment:
Should we not error here? It seems like a valid case in which we should return an error to the user.

Reply:
fixed. this will return an error to the caller now.